File: backend/src/main/java/com/expensetracker/model/Budget.java
package com.expensetracker.model;

import java.math.BigDecimal;

public class Budget {
    private Long id;
    private Long userId;
    private Long categoryId; // null = overall monthly budget
    private String monthYear; // 'YYYY-MM'
    private BigDecimal amount;

    // Joined / calculated fields
    private String categoryName;
    private String categoryIconKey;
    private String categoryColorHex;
    private BigDecimal spent;
    private BigDecimal remaining;
    private Double pctUsed;
    private Integer daysLeft;
    private BigDecimal projected;

    public Budget() {}

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }

    public Long getCategoryId() { return categoryId; }
    public void setCategoryId(Long categoryId) { this.categoryId = categoryId; }

    public String getMonthYear() { return monthYear; }
    public void setMonthYear(String monthYear) { this.monthYear = monthYear; }

    public BigDecimal getAmount() { return amount; }
    public void setAmount(BigDecimal amount) { this.amount = amount; }

    public String getCategoryName() { return categoryName; }
    public void setCategoryName(String categoryName) { this.categoryName = categoryName; }

    public String getCategoryIconKey() { return categoryIconKey; }
    public void setCategoryIconKey(String categoryIconKey) { this.categoryIconKey = categoryIconKey; }

    public String getCategoryColorHex() { return categoryColorHex; }
    public void setCategoryColorHex(String categoryColorHex) { this.categoryColorHex = categoryColorHex; }

    public BigDecimal getSpent() { return spent; }
    public void setSpent(BigDecimal spent) { this.spent = spent; }

    public BigDecimal getRemaining() { return remaining; }
    public void setRemaining(BigDecimal remaining) { this.remaining = remaining; }

    public Double getPctUsed() { return pctUsed; }
    public void setPctUsed(Double pctUsed) { this.pctUsed = pctUsed; }

    public Integer getDaysLeft() { return daysLeft; }
    public void setDaysLeft(Integer daysLeft) { this.daysLeft = daysLeft; }

    public BigDecimal getProjected() { return projected; }
    public void setProjected(BigDecimal projected) { this.projected = projected; }
}

File: app/lib/models/budget.dart
class OverallBudgetModel {
  final double amount;
  final double spent;
  final double remaining;
  final double pctUsed;
  final int daysLeft;
  final double projected;

  OverallBudgetModel({
    required this.amount,
    required this.spent,
    required this.remaining,
    required this.pctUsed,
    required this.daysLeft,
    required this.projected,
  });

  factory OverallBudgetModel.fromJson(Map<String, dynamic> json) {
    return OverallBudgetModel(
      amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
      spent: (json['spent'] as num?)?.toDouble() ?? 0.0,
      remaining: (json['remaining'] as num?)?.toDouble() ?? 0.0,
      pctUsed: (json['pctUsed'] as num?)?.toDouble() ?? 0.0,
      daysLeft: (json['daysLeft'] as num?)?.toInt() ?? 0,
      projected: (json['projected'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class CategoryBudgetModel {
  final int categoryId;
  final String categoryName;
  final String categoryIconKey;
  final String categoryColorHex;
  final double amount;
  final double spent;
  final double remaining;
  final double pctUsed;

  CategoryBudgetModel({
    required this.categoryId,
    required this.categoryName,
    required this.categoryIconKey,
    required this.categoryColorHex,
    required this.amount,
    required this.spent,
    required this.remaining,
    required this.pctUsed,
  });

  factory CategoryBudgetModel.fromJson(Map<String, dynamic> json) {
    return CategoryBudgetModel(
      categoryId: (json['categoryId'] as num).toInt(),
      categoryName: json['categoryName'] as String,
      categoryIconKey: json['categoryIconKey'] as String? ?? 'tag',
      categoryColorHex: json['categoryColorHex'] as String? ?? '#4A90E2',
      amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
      spent: (json['spent'] as num?)?.toDouble() ?? 0.0,
      remaining: (json['remaining'] as num?)?.toDouble() ?? 0.0,
      pctUsed: (json['pctUsed'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class BudgetOverviewModel {
  final OverallBudgetModel overall;
  final List<CategoryBudgetModel> categories;

  BudgetOverviewModel({
    required this.overall,
    required this.categories,
  });

  factory BudgetOverviewModel.fromJson(Map<String, dynamic> json) {
    return BudgetOverviewModel(
      overall: OverallBudgetModel.fromJson(json['overall'] as Map<String, dynamic>),
      categories: (json['categories'] as List<dynamic>?)
              ?.map((e) => CategoryBudgetModel.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}
